src/rgw/rgw_sal.cc

```diff
@@ -97,9 +97,139 @@ RGWObjState::RGWObjState(const RGWObjState& rhs) : obj (rhs.obj) {
   compressed = rhs.compressed;
 }
 
+// to be moved into rgw::sal::rados
+rgw::sal::Driver* DriverManager::read_rados_config(const DoutPrefixProvider* dpp,
+                                              JSONFormattable sal_config,
+                                              rgw::sal::Driver* driver,
+                                              CephContext* cct,
+                                              bool use_gc_thread,
+                                              bool use_lc_thread,
+                                              bool quota_threads,
+                                              bool run_sync_thread,
+                                              bool run_reshard_thread,
+                                              bool run_notification_thread,
+                                              bool use_cache,
+                                              bool use_gc, optional_yield y) {
+
+  driver = newRadosStore();
+  RGWRados* rados = static_cast<rgw::sal::RadosStore* >(driver)->getRados();
```

Comment:
@dang is all of this RadosStore specific initialization supposed to happen right after the driver gets set to a RadosStore? Instead of just passing a dpp, sal_config, and driver to the new functions, I had to also add the other 10 arguments to `init_storage_provider()` for the time being. Should these be passed into the RadosStores json config? Or is there some other way that you had in mind?

Reply:
In principle, I suppose, all this can apply to any store, but the way it's supposed to work is that you have a Factory, and you pass the json into the Factory, and it picks the correct factory instance, which understands all the config in that json.  So you shouldn't need all these config to be passed in, it will live either in the json, or in the ceph.conf (for the RADOS store specific stuff).

Reply:
most of these bool options come from ceph.conf, so can be overridden per-radosgw instance. anything we put in the zone json must apply to all radosgw[-admin]s in the zone